core/src/main/scala/org/bykn/bosatsu/Declaration.scala

```diff
@@ -162,7 +166,8 @@ sealed abstract class Declaration {
           loop(p).map(_ => decl)
         case Var(name) =>
           Expr.Var(None, name, decl)
-        case Match(arg, branches) =>
+        case Match(kind, arg, branches) =>
+          // TODO handle recursive matches
```

Comment:
I think we actually don't need to worry about the recur vs match here. We only need to know that this for doing check that the recursion is allowed. Once we see that it is allowed, it is a normal match.